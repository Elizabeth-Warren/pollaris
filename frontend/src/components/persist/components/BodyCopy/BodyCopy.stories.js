import React from 'react';
import styled from 'styled-components';
import { select, text } from '@storybook/addon-knobs';
import BodyCopy, { BodyCopySizes, BodyCopyWeights } from './index';
import theme from '../../theme';

export default {
  title: 'Components/BodyCopy',
  component: BodyCopy,
  parameters: { jest: ['BodyCopy'] },
};

const sampleText = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec ipsum augue, tempus at felis non, ullamcorper semper nisi. In quis erat lacus. Nunc dapibus risus a purus aliquet, ut maximus diam dapibus. Nunc eu urna ut sem lacinia mattis. Nunc imperdiet nisi bibendum, dapibus orci in, ullamcorper augue. Quisque dictum hendrerit augue, sed congue metus. Proin mattis nec enim et volutpat. Praesent lacinia consectetur nisi, at cursus erat porttitor non.';
export const withKnobs = () => {
  const tag = text('Tag', 'p');
  const copy = text('Text', sampleText);
  const size = select('Size', Object.values(BodyCopySizes), BodyCopy.defaultProps.size);
  const weight = select('Weight', Object.values(BodyCopyWeights), BodyCopy.defaultProps.weight);
  const color = select('Color', Object.keys(theme.colors), BodyCopy.defaultProps.color);
  return (
    <div>
      <BodyCopy
        as={tag}
        size={size}
        weight={weight}
        color={color}
      >
        {copy}
      </BodyCopy>
    </div>
  );
};

export const gallery = () => {
  const copy = text('Text', sampleText);
  return (
    <div>
      {Object.entries(BodyCopySizes).reverse().map(([name, value]) => (
        <BodyCopy as="p" size={value} key={name}>
          {`${name}: `}
          {copy}
        </BodyCopy>
      ))}
    </div>
  );
};

export const withStyledComponent = () => {
  const copy = text('Text', sampleText);
  const size = select('Size', Object.values(BodyCopySizes), BodyCopy.defaultProps.size);
  const StyledCopy = styled.div`
    color: ${theme.colors.red};
  `;
  return (
    <div>
      <BodyCopy
        as={StyledCopy}
        size={size}
      >
        {copy}
      </BodyCopy>
    </div>
  );
};
