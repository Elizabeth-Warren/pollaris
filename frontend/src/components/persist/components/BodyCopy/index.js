import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import theme from '../../theme';

export const BodyCopySizes = {
  '2XS': '2xs',
  XS: 'xs',
  SM: 'sm',
  MD: 'md',
  LG: 'lg',
  XL: 'xl',
};

export const BodyCopyWeights = {
  LIGHT: 'light',
  NORMAL: 'normal',
  BOLD: 'bold',
};

const CopyWrapper = styled.span`
  color: ${({ textColor }) => theme.colors[textColor] || 'inherit'};
  font-family: ${theme.fontFamily.sans};
  font-size: ${({ size }) => {
    switch (size) {
      case BodyCopySizes.XL:
        return theme.fontSize['2xl'];
      case BodyCopySizes.LG:
        return theme.fontSize.xl;
      case BodyCopySizes.SM:
        return theme.fontSize.base;
      case BodyCopySizes.XS:
        return theme.fontSize.sm;
      case BodyCopySizes['2XS']:
        return theme.fontSize.xs;
      case BodyCopySizes.MD:
      default:
        return theme.fontSize.md;
    }
  }};
  font-weight: ${({ weight }) => {
    switch (weight) {
      case BodyCopyWeights.LIGHT:
        return theme.fontWeight.light;
      case BodyCopyWeights.NORMAL:
        return theme.fontWeight.normal;
      case BodyCopyWeights.BOLD:
        return theme.fontWeight.bold;
      default:
        return 'inherit';
    }
  }};
  line-height: ${({ size }) => {
    switch (size) {
      case BodyCopySizes.XL:
        return theme.leading.tight;
      default:
        return theme.leading.snug;
    }
  }};
`;

const BodyCopy = ({
  children, size, as, color, ...other
}) => (
  <CopyWrapper
    as={as}
    size={size}
    textColor={color}
    {...other}
  >
    {children}
  </CopyWrapper>
);

BodyCopy.propTypes = {
  /** Child text or elements */
  children: PropTypes.node.isRequired,
  /** HTML tag name or React element. */
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
  /** Size of the body copy */
  size: PropTypes.oneOf(Object.values(BodyCopySizes)),
  /** Font weight of the body copy */
  weight: PropTypes.oneOf(['inherit', ...Object.values(BodyCopyWeights)]),
  /** Selected Color applied directly from theme */
  color: PropTypes.oneOf(['inherit', ...Object.keys(theme.colors)]),
};

BodyCopy.defaultProps = {
  size: BodyCopySizes.MD,
  weight: 'inherit',
  color: 'inherit',
};

export default BodyCopy;
