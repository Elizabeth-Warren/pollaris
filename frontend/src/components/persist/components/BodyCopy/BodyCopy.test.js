import React from 'react';
import renderer from 'react-test-renderer';
import BodyCopy, { BodyCopySizes } from './index';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <div>
        {Object.entries(BodyCopySizes).reverse().map(([name, value]) => (
          <BodyCopy as="p" size={value} key={name}>
            {`${name}: `}
            Hello world
          </BodyCopy>
        ))}
      </div>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
