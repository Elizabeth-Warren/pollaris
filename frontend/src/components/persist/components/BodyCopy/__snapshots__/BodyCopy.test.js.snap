// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  color: inherit;
  font-family: 'Ringside Regular A','Ringside Regular B','Helvetica',sans-serif;
  font-size: 36px;
  font-weight: inherit;
  line-height: 1.17;
}

.c1 {
  color: inherit;
  font-family: 'Ringside Regular A','Ringside Regular B','Helvetica',sans-serif;
  font-size: 26px;
  font-weight: inherit;
  line-height: 1.3;
}

.c2 {
  color: inherit;
  font-family: 'Ringside Regular A','Ringside Regular B','Helvetica',sans-serif;
  font-size: 18px;
  font-weight: inherit;
  line-height: 1.3;
}

.c3 {
  color: inherit;
  font-family: 'Ringside Regular A','Ringside Regular B','Helvetica',sans-serif;
  font-size: 16px;
  font-weight: inherit;
  line-height: 1.3;
}

.c4 {
  color: inherit;
  font-family: 'Ringside Regular A','Ringside Regular B','Helvetica',sans-serif;
  font-size: 14px;
  font-weight: inherit;
  line-height: 1.3;
}

.c5 {
  color: inherit;
  font-family: 'Ringside Regular A','Ringside Regular B','Helvetica',sans-serif;
  font-size: 12px;
  font-weight: inherit;
  line-height: 1.3;
}

<div>
  <p
    className="c0"
    size="xl"
  >
    XL: 
    Hello world
  </p>
  <p
    className="c1"
    size="lg"
  >
    LG: 
    Hello world
  </p>
  <p
    className="c2"
    size="md"
  >
    MD: 
    Hello world
  </p>
  <p
    className="c3"
    size="sm"
  >
    SM: 
    Hello world
  </p>
  <p
    className="c4"
    size="xs"
  >
    XS: 
    Hello world
  </p>
  <p
    className="c5"
    size="2xs"
  >
    2XS: 
    Hello world
  </p>
</div>
`;
